lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -217,25 +247,85 @@ actor ConnectorSink is Sink
+      Fail()
+    end
+
     var receive_ts: U64 = 0
     ifdef "detailed-metrics" then
-      receive_ts = WallClock.nanoseconds()
+      receive_ts = Time.nanos()
       _metrics_reporter.step_metric(metric_name, "Before receive at sink",
         9998, latest_ts, receive_ts)
     end
 
-    ifdef "trace" then
-      @printf[I32]("Rcvd msg at ConnectorSink\n".cstring())
-    end
     try
-      let encoded = _encoder.encode[D](data, _wb)?
+      if _notify.credits == 0 then
+        // TODO: add token management back to connector sink protocol?
+        // Ideas?
+        // 1. We can't simply use _apply_backpressure because the socket
+        //    probably is writeable, which would mean ASIO would immediately
+        //    trigger a writeable event, and we'd have to check credits
+        //    and defer again, causing an ASIO busy wait loop??
+        //    Maybe that's OK?  But if the remote were to crash without
+        //    sending an Ack with credit replenishing, then we could
+        //    busy wait for a long time.
+        // 2. If we have zero credits, then we set
+        //    pony_asio_event_set_writeable(false), then we also apply
+        //    backpressure *without* also resubscribing
+        //    pony_asio_event_set_writeable(true)!  We want to mute
+        //    upstreams, and the rest of _apply_backpressure() will do
+        //    that for us.
+        //    * For events that arrive here while pony**writeable(false),
+        //      can we simply call _writev() and then know exactly how
+        //      many credits correspond to stuff in _pending??
+        //      * Do we change the 2-tuple in _pending to a 3-tuple, where
+        //        the 3rd element is the # of credits that correspond?
+        //        * Tricksy, because a single credit of message may be more
+        //          than one writev array item!
+        None
+      else
+        _notify.credits = _notify.credits - 1
+      end
+      let encoded1 = _encoder.encode[D](data, _wb)?
+      var encoded1_len: USize = 0
+      for x in encoded1.values() do
+        encoded1_len = encoded1_len + x.size()
+      end
+      // We do not include MessageMsg size overhead in our offset accounting
+      _twopc_current_offset = _twopc_current_offset + encoded1_len
+
+      let encoded2 =
+        try
+          let msg = _notify.make_message(encoded1)?
+          let w1: Writer = w1.create()  
+          let w2: Writer = w2.create()
+          let b = cp.Frame.encode(msg, w1)
+          w2.u32_be(b.size().u32())
+          w2.write(b)
+          w2.done()
+        else
+          Fail()
+          encoded1
+        end
 
       let next_seq_id = (_seq_id = _seq_id + 1)
-      _writev(encoded, next_seq_id)
+      _writev(encoded2, next_seq_id)
 
-      // TODO: Should happen when tracking info comes back from writev as
-      // being done.
-      let end_ts = WallClock.nanoseconds()
+      let end_ts = Time.nanos()
```

Comment:
Same question.  There was a commit where Simon updated all these to WallClock.nanoseconds().